llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -14903,12 +14903,31 @@ SDValue DAGCombiner::reduceLoadWidth(SDNode *N) {
   AddToWorklist(NewPtr.getNode());
 
   SDValue Load;
-  if (ExtType == ISD::NON_EXTLOAD)
-    Load = DAG.getLoad(VT, DL, LN0->getChain(), NewPtr,
-                       LN0->getPointerInfo().getWithOffset(PtrOff),
-                       LN0->getOriginalAlign(),
-                       LN0->getMemOperand()->getFlags(), LN0->getAAInfo());
-  else
+  if (ExtType == ISD::NON_EXTLOAD) {
+    const MDNode *OldRanges = LN0->getRanges();
+    const MDNode *NewRanges = nullptr;
+    /* If LSBs are loaded and the truncated ConstantRange for the OldRanges
+       metadata is not the full-set for the NewWidth then create a NewRanges
+       metadata for the truncated load */
+    if (ShAmt == 0 && OldRanges) {
+      Type *NewTy = VT.getTypeForEVT(*DAG.getContext());
+      const unsigned NewWidth = NewTy->getIntegerBitWidth();
+      const ConstantRange CR = getConstantRangeFromMetadata(*OldRanges);
+      const ConstantRange TruncatedCR = CR.truncate(NewWidth);
+
+      if (!TruncatedCR.isFullSet()) {
+        Metadata *Bounds[2] = {ConstantAsMetadata::get(ConstantInt::get(
+                                   NewTy, TruncatedCR.getLower())),
+                               ConstantAsMetadata::get(ConstantInt::get(
+                                   NewTy, TruncatedCR.getUpper()))};
```

Comment:
You can pass LLVMContext instead of NewTy to these. Then you won't need the getTypeForEVT lookup.

Reply:
Implement without using getTypeForEVT.